### PR TITLE
[REVIEW] Centralize logic for getattr->getitem override

### PR DESCRIPTION
**Pull request by vyasr**
There are a few classes that need `obj.foo` to behave like `obj['foo']`. They were implementing this independently, but getting it just right can be tricky, so this centralizes that logic into a single mixin.

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -638,34 +639,25 @@ def __dir__(self):
...
-        else:
-            if key in self:
-                return self[key]
-
-        raise AttributeError("'DataFrame' object has no attribute %r" % key)
+            if key not in self._PROTECTED_KEYS:
+                try:
+                    # If a column already exists, set it.
+                    self[key]  # __getitem__ to verify key exists
+                    self[key] = col
```
Comment (isVoid) [inline python/cudf/cudf/core/dataframe.py:654]: Not part of this PR but would it be better to use `self._data[key] = col` here? Looks like a special case we can bypass `__getitem__` overhead.

File: python/cudf/cudf/utils/utils.py [hunk 2]
```diff
@@ -280,6 +281,33 @@ def __get__(self, instance, cls):
...
+
+    Subclasses may define a `_PROTECTED_KEYS` set as a class attribute to
+    indicate any keys that may be accessed by __getitem__. Testing for these
+    keys allows this class to prevent recursion errors when subclasses are
+    copied (see below for more details). Note that a cleaner solution for this
+    problem would be to override one of the serialization protocols (e.g.
+    `__reduce__` or `__setstate__`, but this class may be used in complex
+    multiple inheritance hierarchies that might also override serialization.
+    The solution here is a minimally invasive change to simplify such patterns.
+    """
```
Comment (isVoid) [inline python/cudf/cudf/utils/utils.py:295]: Great explanation, though I think some of the comments can be moved to the PR description page. For documentation purpose, this may reveal too much implementation (plus non-implementation) detail. The mentioning of "recursion prevention" is duplicate of that below.

I would, however, like to know more about the `__reduce__` and `__setstate__` solution. Could you explain how that might be done?

**Review by isVoid — COMMENTED** [review]
Generally a big fan of making codes share more common logic and making use of OOD patterns. A few minor comments below.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:654]: No, unfortunately not. This code is a bit more general this way because it also works in cases where `self` does not have an `_data` attribute. In particular, it's being used by `Rolling`.

File: python/cudf/cudf/utils/utils.py (hunk 2 above)
Comment (vyasr) [reply python/cudf/cudf/utils/utils.py:295]: Good points. This class is entirely internal and the docstring is intended to be for developers, so some amount of implementation detail is intentional, but I'm not sure what the right balance is. What do you think?

Comment (vyasr) [reply python/cudf/cudf/utils/utils.py:295]: Your `__setstate__` question is a much more complicated one.

Short version: you can change how an object is copied (via `copy.copy`) and serialized (`via pickle.dump`) using `__setstate__`, so if you implement those you can implicitly bypass the problem that the `_PROTECTED_KEYS` check solves. The problem only happens in the default copying implementation, so the problematic code in the CPython `copy` implementation is never executed if you override other methods that change the copying process.

Long version: The default behavior when you do `copy.copy(obj)` is basically something like `new_obj = type(obj).__new__(); for k in obj.__dict__: getattr(new_obj, k)`. However, Python objects offer multiple different hooks that the user can implement to customize how the object is copied. Therefore, when `copy.copy(obj)` is invoked, before the `for` loop I wrote above is run Python has to introspect `obj` for the presence of these other methods, which would indicate that copying should proceed by calling those rather than the default copying of `__dict__`. The recursion problem arises because looking up whether these methods are defined eventually leads to calling `__getattr__` on the object after it is created with `__new__`; since `__init__` is never called, the special `_data` attribute (for example) that would normally exist on the object when created via normal means is not defined, so the object is in an invalid state where `__getattr__` recursively searches for `self._data`. 

You can avoid this problem if you actually define one (or more) of those special methods. Python has a pretty complex and poorly documented hierarchy of different functions that can be used to modify the way that both copying and pickling of an object works; it turns out that many (but not all) functions are shared between these protocols. The only place I know to find the full dispatch sequence is by actually inspecting the CPython source code. Here's a rough summary of the important points, though. To customize the behavior of `copy`, the first thing that is used is the `__copy__` method. If it exists, then that's all that happens. If that is not defined, then `__getstate__` is invoked (if it exists) on the object being copied to get some data representing the "state" of the object. Then, once the new object is created (via `__new__`, with `__init__` never called), the new object's `__setstate__` method is called with the state generated by `__getstate__`. If neither `__getstate__` nor `__setstate__` is defined, the default copying of `__dict__` happens.

All of the methods I mentioned above (and a few others like `__getnewargs_ex__`) are part of Python's copying protocol, which is fundamentally implemented by another dunder method `__reduce__` that calls `__copy__` and `__getstate__` in the order I described above. You can instead override `__reduce__` itself to completely customize the process, but its signature is pretty cumbersome. That method must return a tuple that contains, at minimum, the class constructor to call (e.g. `cudf.DataFrame`), along with the arguments necessary to reconstruct this object. However, that tuple can also contain other items that allow you to customize things like setting the state of the object after initialization if there is state that cannot be controlled via the constructor alone. It's a pretty heavy hammer, though, and I'd avoid it unless absolutely necessary.

In this case, I would view any of these solutions as somewhat fragile because we're using this mixin as part of an inheritance tree with classes (like `Serializable`) that also may modify these methods, so it seems inadvisable to rely on subclasses not to override this behavior in problematic ways. As a result, the `_PROTECTED_KEYS` approach seems like the safer choice.

Hopefully that answers everything you could ever want to know about Python copying :)

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:654]: Oh whoops, I missed that this was in `__setitem__`, not the shared `__getitem__`. Yes, I think that would work, let me give it a shot.

File: python/cudf/cudf/utils/utils.py (hunk 2 above)
Comment (isVoid) [reply python/cudf/cudf/utils/utils.py:294]: This explanation of python copy mechanism is super thorough. Thank you for this. I have to admit the first time I read the docstring I thought it discusses two different issues. But now it makes more sense. 

So to put it in perspective, if user calls `copy.copy` for one of cuDF object, it may cause a recursion error. Because in the new object, `__init__` was not invoked and `_PROTECTED_KEYS` is not yet defined, so `__getattr__` is called again.

However, cuDF/Pandas users are more advised to use the defined `copy` method on these objects, such as `Series.copy`, `DataFrame.copy` to perform copy, since these objects has underlying buffer that requires custom memory handling. It would be nice to see how we can adapt them into standard Python copy protocol - but an obvious counter argument is, if Python copy protocol is not well documented, perhaps we should discourage user from using `copy.copy` explicitly on cuDF objects? If the protocol is not well documented, it's not part of the language spec but implementation detail and should not be dependent on.

Besides top level objects that requires `GetAttrGetItemMixin`, is it possible there are internal objects that requires this as well? How should we consider python copy protocol to them? My personal take is we should discourage use of `copy.copy` unless deemed perfectly safe to do so, (against trivial objects, for example).

Comment (isVoid) [reply python/cudf/cudf/utils/utils.py:294]: Re docstring: personally I'm happy to read detailed design level ideas laid out in docstrings. Very often the issue is not detailed enough :). But to me what's in the docstring now seems
1. Either too succinct and contains some unstated assumptions (`that might also override serialization`, this happens when subclasses also overrides the serialization protocol, but not accounting for all parent classes when doing so. But why?). 
2. Or too detailed to account for the fact of `recursion error prevention`.

Comment (vyasr) [reply python/cudf/cudf/utils/utils.py:294]: I've reread the docstring and I definitely agree that I could do a better job of separating implementation details from what inheriting classes need to know in order to use it properly, which I think is the most important dividing line for developer documentation. I've tried to address this in the most recent rewrite, where I focus the docstring on telling developers what they need to do to safely use this class, and added a longer inline comment inside the class explaining the details of why this fix is necessary. Let me know if you think this is better.

Comment (vyasr) [reply python/cudf/cudf/utils/utils.py:294]: On the topic of whether or not `copy.copy` is safe, the expected behavior of `copy` and `pickle` methods is well-defined by the standard. Furthermore, the different ways to override these processes I mentioned above are more than just an implementation detail, they are actually promised based on [the various PEPs mentioned here](https://docs.python.org/3/library/pickle.html#data-stream-format). The problem is that none of these behaviors have good documentation. If you write a class that needs to customize the behavior of copying, deep copying, or pickling, the only good sources of information I know of for determining the best way to do this are the various PEPs linked in the pickle documentation linked above, or the CPython source code itself. In other words, this doesn't have any impact on users, but it can make developers lives very hard :) The fact that redefining other dunder methods like `__getattr__` can have unexpected effects on these protocols is I think just a fact of life that we have to live with.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py]: I've modified both the above lines to take faster paths. I've tried this and my tests pass locally, but I'm not 100% sure that the set is exactly equivalent. I'm reasonably confident that replacing `self[key]` with `self._data._select_by_label_grouped` is completely equivalent here based on the possible values of `key` in this context, but `DataFrame.__setitem__` even for scalar keys seems a bit more complex and I'm a bit concerned that there might be some kinds of inputs for which this won't work. Maybe @shwina would know for sure?

File: python/cudf/cudf/utils/utils.py (hunk 2 above)
Comment (isVoid) [reply python/cudf/cudf/utils/utils.py]: Re docstring: it looks very nice to me now. Thanks more making the rewrite!

File: python/cudf/cudf/core/tools/datetimes.py [hunk 3]
```diff
@@ -479,7 +479,10 @@ def __init__(self, n=1, normalize=False, **kwds):
                 f"Keyword arguments '{','.join(list(wrong_kwargs))}'"
                 " are not yet supported in cuDF DateOffsets"
             )
-        self._scalars = _DateOffsetScalars(scalars)
+        # pandas DateOffset objects are made immutable by disabling
+        # `__setattr__`, but this class adds the _scalars attribute (also
+        # immutable after construction), so we bypass `DateOffset.__setattr__`.
+        object.__setattr__(self, "_scalars", _DateOffsetScalars(scalars))
```
Comment (shwina) [inline python/cudf/cudf/core/tools/datetimes.py:485]: FWIW, we _don't_ disable `setattr` for `DateOffset` - so do we need this?

Comment (vyasr) [reply python/cudf/cudf/core/tools/datetimes.py:485]: `cudf.DateOffset` inherits from `pandas.DateOffset`, which does disable `__setattr__`. Previously, this was being overcome by defining `__setattr__` in this class to only allow setting of `_DateOffsetScalars` (as values, not as keys), which wasn't a very well targeted solution (you could do `obj.any_attribute_name = _DateOffsetScalars(...)` and it would work). Now I just defer to the parent implementation of disabling `__setattr__` and bypass it here.

Comment (shwina) [reply python/cudf/cudf/core/tools/datetimes.py:485]: Ah right. Sorry, I thought our PR here was merged, where we remove the inheritance: https://github.com/rapidsai/cudf/pull/7078

Comment (vyasr) [reply python/cudf/cudf/core/tools/datetimes.py:485]: Got it, was not aware of that PR. Depending on which PR gets merged first, we can remove this block and just do `self._scalars = _DateOffsetScalars(scalars)`. I'll make a note on that PR too.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py]: Discussed with @shwina offline. This is _probably_ safe, but there are enough edge cases that it's hard to be entirely sure. I'm going to change this back for now but keep it in mind in case this ever shows up in profiles.

File: python/cudf/cudf/utils/utils.py [hunk 4]
```diff
@@ -279,6 +280,42 @@ def __get__(self, instance, cls):
...
+    # solution here is a minimally invasive change that avoids such conflicts.
+    _PROTECTED_KEYS: Union[FrozenSet[str], Set[str]] = frozenset()
+
+    def __getattr__(self, key):
+        if key in self._PROTECTED_KEYS:
+            raise AttributeError
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(f"{type(self)} object has no attribute {key}")
```
Comment (shwina) [inline python/cudf/cudf/utils/utils.py:316]: This should give slightly neater results:

```suggestion
            raise AttributeError(f"{type(self).__name__} object has no attribute {key}")
```

**Review by shwina — APPROVED** [review]
Looks great. Tiny suggestion that you can feel free to ignore

**Comment by vyasr** [comment]
> Looks great. Tiny suggestion that you can feel free to ignore

Thanks, went ahead an accepted. Not like CI is running right now anyway.

**Review by isVoid — APPROVED** [review]

**Comment by shwina** [comment]
@gpucibot merge
